Add events to Broadlink after learning a code

zineer opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
When a code is learned by the Broadlink remote component, there isn't a simple way to get the learned code back into an external system. I added events being sent when a code is learnt (and events when an error occurs). I created new events based on the `remote` component (`remote_event`) so that this idea can be expanded to other `remote` components, or additional event types added.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16615

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

homeassistant commented:
Hi @zineer, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#46645).

Once you do that we will be able to review and accept this pull request.

Thanks!

dgomes commented:
@felipediel could you have a look ?

felipediel commented:
It sounds like a great idea!

Two suggestions.
1. Create a separate event for each type to make it more flexibe in case we want to add events with different payloads, eg:
```python3
self.hass.bus.async_fire(
    RM_EVENT_CODE_RECEIVED,
    {
        "device_id": self.unique_id,
        "command": command,
        "code": code,
    },
)
```
```python3
self.hass.bus.async_fire(
    RM_EVENT_ANOTHER_EVENT,
    {
        "device_id": self.unique_id,
        "something_else": "Hooray!",
    },
)
```
2. The learning process have multiple steps, maybe we can add "enter learning" and "sweep frequency" events to help the user press the button at the right time.
```python3
self.hass.bus.async_fire(
    RM_EVENT_ENTER_LEARNING,
    {
        "device_id": self.unique_id,
        "error": "Failed to enter learning mode",  # If there is no error, it worked.
    },
)
```
Just dropping some ideas. The code looks good in general.

felipediel commented:
/still valid

I would reinforce the first suggestion in order to avoid breaking changes later.

zineer commented:
Still active. Hoping to get to the suggested changes soon.

felipediel commented:
/still valid

zineer commented:
Still active.

emontnemery commented:
@zineer can you rebase the PR please?

emontnemery commented on homeassistant/components/remote/__init__.py:
```diff
@@ -57,6 +57,9 @@
 SUPPORT_LEARN_COMMAND = 1
 SUPPORT_DELETE_COMMAND = 2
 
+EVENT_TYPE = "remote_event"
```
This should the prefixed with the integration's domain
```suggestion
EVENT_TYPE = "broadlink_remote_event"
```

zineer replied:
@emontnemery I added it to the `remote` component with the idea that other `remote` components would be able to implement the event and have a common set of base events that would be used. If there isn't any interest in that, I could move it into the `broadlink` component and prefix as suggested.

emontnemery replied:
If this should indeed live in the `remove` integration, I think the event's data should have a common base with optional  integration specific data.

zineer committed:
Add events to Broadlink after learning a code

zineer committed:
Updates to different event types per command

zineer commented:
@felipediel @emontnemery I finally fit in some time for this. I have rebased onto latest dev and changed the event type as suggested. 
(On a semi-related note and probably not one you guys control, but could be mentioned to someone, the [Event docs](https://developers.home-assistant.io/docs/dev_101_events) don't really specify the best practices for doing new event types or naming, etc. I think I based my original single `EVENT_TYPE` with a `type` property on the events I observed from other components, but I don't recall which exactly.)

emontnemery commented on homeassistant/components/broadlink/remote.py:
```diff
@@ -282,9 +283,16 @@ async def async_learn_command(self, **kwargs):
+            error_message = f"{service} canceled: {self.entity_id} entity is turned off"
+
+            self.hass.bus.async_fire(
+                RM_EVENT_LEARNED_COMMAND,
```
Will it be easier to build automations around this if the error cases have their own event type:
```suggestion
                RM_EVENT_LEARN_COMMAND_FAILED,
```

zineer replied:
Good idea. Have updated with the new failed event.

zineer committed:
separate error event to own event type
